src/NServiceBus.Core/MessageInterfaces/MessageMapper/Reflection/MessageMapper.cs

```diff
@@ -250,6 +259,8 @@ static string GetTypeName(Type t)
 
         readonly object messageInitializationLock = new object();
 
+        HashSet<Type> initializedTypes = new HashSet<Type>();
```

Comment:
@bording made me understand that the non concurrent collections can actually corrupt internal state if used concurrently so I will switch to a concurrent dictionary (marked the PR as WIP)

@SimonCropp Me an @bording couldn't find a way to refactor this to avoid this new dictionary. I propose that we proceed with this PR and performs the refactoring (we can perhaps pair up?) as a separate PR to not delay this?